feature/sniffs_system - twigcs

adrienrn opened the pull request:
Here's `twigcs` - Twig Code Sniffer - which intend to be a "sniff engine" to check more thoroughly twig templates including whitespaces, trailing commas or missing parameters to functions/filters, etc.

If it sounds familiar, it's because it's heavily inspired on [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer).

I tried to not change anything to the current `lint` command. _Everything should work as it was before._

## Internals

There's two kind of sniffs:

- **Pre-parser:** these are sniffs for formatting mainly, whitespace, blank lines, trailing comma, etc. They work using a custom tokenizer (a custom lexer), see "Why?" below.
- **Post-parser:** these are "higher-order" sniffs, for functions, filters, check arguments value, usage, deprecation notices, etc. They work using Twig node visitors.

You define a set of rules (`src/Twig/Lint/Ruleset.php`) with options in a `twigcs.yml` file at the root of your twig project.

```yml
ruleset:
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\DumpSniff'
    options:
      severity: 3
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\WhitespaceBeforeAfterExpression'
    # no options means default severity eg. = 5
```

Once loaded, the Linter class (`src/Twig/Lint/Linter.php`) will process every requested files against that set of rules and produce a Report object (`src/Twig/Lint/Report.php`) with rule violations, context and statistics.

Running `twigcs` against the test fixtures files would produce:

```
./bin/twig-lint twigcs --working-dir tests/Asm89/Twig/Lint/Test/Fixtures/config ./tests/Asm89/Twig/Lint/Test/Fixtures/ --exclude hash_4.twig
```

![image](https://user-images.githubusercontent.com/5598481/32692626-be20b996-c71b-11e7-8bff-fe1dd01420dc.png)

Check the many _PHPUnit_ tests I've made to dig further!

## Todos
- [x] Sniffs via node visitors (post parser)
- [x] Pre parser sniffs via a custom Tokenizer (that handles whitespace, comment, and EOL tokens)
- [x] Base sniffs (Standards/Generic)
- [x] Linter main class
- [x] Load sniffs from config file
- [ ] Error / Exception handling
- [x] Increase log-level
- [x] Stress test with 100~ templates and all sniffs (forking process?)
- [x] Add to `twigcs.yml`:
  - paths / files to process
  - exclude files / patterns
- [x] Allow to load sniffs from another directory (autoload, etc)
- [x] Handle deprecation notices from twig using `set_error_handler`. The StubbedCore makes it difficult but https://github.com/asm89/twig-lint/pull/24 should make it easier.
- [x] Severity option and filtering when displaying the report (or better, processing the sniffs).
- [ ] Have a "Standards" feature, to make the definition easier. Now it's fine but the more sniff you'll have, the more `twigcs.yml` will be a clutter; and it would be nice to have a "standard" syntax like PSR-2 for twig, share by everyone.
- [ ] Accept the template via stdin.
- [ ] Documenting Tokenizer.php a bit more.
- [ ] See if I can improve performance for the tokenizer. There's no performance problems yet, but  (this article looks great!)[https://nikic.github.io/2011/10/23/Improving-lexing-performance-in-PHP.html].

## Why?

In the end, I did it to integrate this to my CI process, using [PHPstan](https://github.com/phpro/grumphp/pull/199) or Codacy. 

All of my many twig projects are maintained by many people (including freelance developers) and with the activity growing, I don't have time anymore to spend that much time reading pull requests 😔.

I created my own tokenizer / preprocessor out of the official lexer because it was not possible to detect space and all punctuations otherwise, see those issues:

- https://github.com/twigphp/Twig/issues/1880
- https://github.com/twigphp/Twig/issues/2475
- https://github.com/twigphp/Twig/issues/2451

## Final notes!

If it's too much, I'll continue this on my fork or create a separate repository. :)

adrienrn committed:
Add phpunit/phpunit to composer.json

adrienrn committed:
First attempt at sniffs

adrienrn committed:
Try to organize things a bit

adrienrn committed:
Fixing some things, adding one warning for trans

adrienrn committed:
Some more sniffs + default severity and utils

adrienrn committed:
Preprocessor (wip)

adrienrn committed:
Clean preprocessor + Pre / Post parser sniff interface and abstract classes

adrienrn committed:
Glue things together, finally (Linter.php + Ruleset.php)

adrienrn committed:
New Report class, enable/disbale sniff, Linter makes more, clean tests.. renaming some things

adrienrn committed:
Detect simple / double quotes (SimpleQuotesSniff.php) + clean

adrienrn committed:
Add position & filename to the token / node and to the final report (wip)

adrienrn committed:
Compute stats on the final report + tests

adrienrn committed:
EnforceHashKeyQuotesSniff + tests

adrienrn committed:
Sniffs for hashes + tests

adrienrn committed:
Return a Twig_Node from CatchAll to be able to sniff them!

adrienrn committed:
Clean some things (tests mainly)

adrienrn committed:
Load files using Finder, first try at twigcs.yml + add new sniff for DisallowTabIndentSniff + refactoring for tests

adrienrn committed:
Added twigcs.yml - loading, parsing, validating + tests with Ruleset.php

adrienrn committed:
Reorganize tests + fix comments with {% and {{ inside

adrienrn committed:
Reorganize tests (fixtures files)

adrienrn committed:
Factorization of code (AbstractSniff) + severity as an option

adrienrn committed:
Added RulesetFactory to easy things with config files

adrienrn committed:
New twigcs command + various fixes

adrienrn committed:
Replace simple array by SniffViolation.php + fixes

adrienrn committed:
Use \Twig_Source + some fixes

adrienrn committed:
Add copyright on all new files!

adrienrn committed:
PHPdoc (almost) everywhere

adrienrn committed:
Adjust composer.json version for php5.3 - php5.5

adrienrn committed:
Fixes for php5.3+ short array and const

adrienrn committed:
Make Travis work with php5.3

adrienrn committed:
Another try at fixing php5.3 in travis

adrienrn committed:
Tokenizer errors + catch in Linter.php with new message

adrienrn committed:
Add missing *Sniff suffix to WhitespaceBeforeAfterExpression.php

adrienrn committed:
Split sniffs to do one specific task + more tests

adrienrn committed:
Remove the use of Twig_TokenParserBroker (deprecated) + fix all deprecation warnings - #31

Conflicts:
	src/Asm89/Twig/Lint/StubbedEnvironment.php

adrienrn committed:
Remove the use of Twig_TokenParserBroker (deprecated) + fix all deprecation warnings - #31

Conflicts:
	src/Asm89/Twig/Lint/StubbedEnvironment.php

Conflicts:
	src/Asm89/Twig/Lint/StubbedEnvironment.php

adrienrn committed:
paths / exclude via twigcs.yml + fixes for files, deprecated warnings

adrienrn committed:
Fix EnsureBlankAtEOFSniff when file is empty

adrienrn committed:
Added text formatter + filtering on level or severity

adrienrn committed:
Fix EnsureHashKeyQuotesSniff for complex keys eg. expressions

asm89 commented:
@adrienrn Woa, this is pretty cool! I need to properly sit down to go through this though. :)

adrienrn committed:
Timing / memory consumption + 'explain' options on TextFormatter

adrienrn committed:
Fix EnsureHashKeyQuotesSniff again + fix tests

adrienrn committed:
phpcs / php compat

adrienrn committed:
Apply fix for travis, php 5.3 and trusty

adrienrn committed:
DisallowCommentedCodesniff + tests

adrienrn committed:
Added an autoloader and a config for external sniffs

adrienrn committed:
Deleted unused fixture file 'twigcs.yml'

adrienrn committed:
Find and load 'twigcs_global.yml' from $HOME for external sniffs

adrienrn committed:
Remove typo in class name (case sensitivity)

adrienrn commented:
Update:

Added some major things (more powerful config, external sniffs) and some minor ones (time / memory metrics, better output).

An example of the twigcs.yml of a project could be:

```yml
paths:
  - 'Resources/views'
exclude:
  - 'node_modules'
  - 'Resources/public'
ruleset:
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\DeprecatedTemplateNotationSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\DisallowCommentedCodeSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\DisallowDumpSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\DisallowIncludeTagSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\DisallowTabIndentSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\EnsureBlankAtEOFSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\EnsureHashKeyQuotesSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\EnsureHashSpacingSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\EnsureHashTrailingCommaSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\EnsureQuotesStyleSniff'
    options:
      style: 'TYPE_SINGLE_QUOTES'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\EnsureTranslationArgumentsSniff'
  -
    class: 'Asm89\Twig\Lint\Standards\Generic\Sniffs\EnsureWhitespaceExpressionSniff'
```

![image](https://user-images.githubusercontent.com/5598481/33349520-c073316a-d49a-11e7-85d2-688d53fec901.png)

![image](https://user-images.githubusercontent.com/5598481/33349637-45675d74-d49b-11e7-8ac1-1a1fc5afd8e0.png)

I also added the possibility to load custom sniffs; using the `standardPaths` configuration key. 

You can add it to your `twigcs.yml` but it makes more sense to create a file `~/.twigcs/twigcs_global`. Everything in that file will be merged with the one from your project.

```yml
ruleset: []
standardPaths:
  '\Acme\TwigCS':
    - '/home/vagrant/work/acme-twigcs/src'
```

Working on some proper "standard" next to easy things up.

adrienrn committed:
Fix two words tests 'same as' and 'divisible by' for example

adrienrn committed:
Forward-compatibility for phpunit and PHP7

adrienrn committed:
Merge branch 'remove/deprecated_warnings' into feature/sniffs_system

Conflicts:
	src/Asm89/Twig/Lint/Command/LintCommand.php
	src/Asm89/Twig/Lint/StubbedEnvironment.php
	tests/Asm89/Twig/Lint/Test/StubbedEnvironmentTest.php

adrienrn committed:
Oops composer.json after merge

seyfer commented:
@adrienrn see my comment here
https://github.com/asm89/twig-lint/pull/36

you better make your repo https://github.com/adrienrn/twig-lint
as a continuation :) then people can merge their PRs in your repo and you will maintain it, if you want
